File: 2020/day7/python/day7.py
#!/usr/bin/env python3
from os.path import dirname
from os.path import realpath
from os.path import join
import time
import pytest
import re
from collections import defaultdict


def transform_input(input_):
    # custom transform for the day

    bag_contains = defaultdict(list)

    for sentence in input_.splitlines():
        bag, contents = sentence.split("contain")
        bag_color = re.sub("bag(.*)", "", bag).strip()
        bag_contents = [re.sub("bag.*", "", c) for c in contents.split(",")]

        for parts in bag_contents:
            # Include bag if we have something in it
            match = re.search("^\\s+(\\d+)", parts)
            if match:
                n_bags = match.group(0)
                color = parts[len(n_bags) :].strip()
                bag_contains[bag_color].append([color, int(n_bags)])

    return bag_contains


def read_input(file_name="../data/input.txt"):
    dir_path = dirname(realpath(__file__))
    with open(join(dir_path, file_name), "r") as f:
        input_ = transform_input(f.read())

    return input_


def contains_bag(input_, color):
    if color not in input_.keys():
        return False
    return any(
        [
            True if c == "shiny gold" else contains_bag(input_, c)
            for c, _ in input_[color]
        ]
    )


def solve_part1(input_):
    # Go through each unique key and see if it leads to a shiny gold bag
    return sum([contains_bag(input_, c) for c in input_])


def bags_in_bag(bag, color, n_bags):
    if not bag[color]:
        return n_bags
    bags_in_color = 0
    for c, n in bag[color]:
        bags_in_color += n_bags * bags_in_bag(bag, c, n)
    return n_bags + bags_in_color


def solve_part2(input_):
    # Subtract 1 - don't count the "shiny gold" bag itself
    return bags_in_bag(input_, "shiny gold", 1) - 1


def main():
    input_ = read_input()

    t0 = time.time()
    part1 = solve_part1(input_)
    time_part1 = round((time.time() - t0) * 1e3)
    print(f"Solution to part one: {part1} (time taken {time_part1}[ms])")

    t0 = time.time()
    part2 = solve_part2(input_)
    time_part2 = round((time.time() - t0) * 1e3)
    print(f"Solution to part one: {part2} (time taken {time_part2}[ms])")


if __name__ == "__main__":
    main()


@pytest.mark.parametrize(
    "input1, output1",
    [("../data/test_input0.txt", 4), ("../data/test_input2.txt", 4)],
)
def test1(input1, output1):
    assert solve_part1(read_input(input1)) == output1


@pytest.mark.parametrize(
    "input2, output2",
    [("../data/test_input0.txt", 32), ("../data/test_input1.txt", 126)],
)
def test2(input2, output2):
    assert solve_part2(read_input(input2)) == output2

File: 2020/day7/cpp/src/day7.cpp
#include "2020/day7/cpp/include/day7.hpp"

#include <boost/algorithm/string.hpp>
#include <regex>

#include "utils/cpp/include/utils.hpp"

using day7::bag_contents;
using day7::input_type;

input_type day7::transform_input(std::vector<std::string> &input) {
  input_type bag_contains;

  for (auto &sentence : input) {
    std::vector<std::string> split;
    utils::split_string(sentence, "contain", split);
    auto bag = split[0], contents = split[1];

    auto bag_color = std::regex_replace(bag, std::regex("bag(.*)"), "");
    boost::algorithm::trim(bag_color);
    std::vector<std::string> parts;
    utils::split_string(contents, ',', parts);

    std::regex rgx("^\\s+(\\d+)");
    for (auto &part : parts) {
      part = std::regex_replace(part, std::regex("bag.*"), "");
      std::smatch match;
      if (std::regex_search(part, match, rgx)) {
        auto n_bags = match[0];
        auto color =
            part.substr(n_bags.length(), part.length() - n_bags.length());
        boost::algorithm::trim(color);
        bag_contains[bag_color].push_back(
            std::make_pair(color, std::stoi(n_bags)));
      }
    }
  }
  return bag_contains;
}

input_type day7::read_input(std::string file_name) {
  auto input = utils::read_file(file_name);
  return transform_input(input);
}

bool day7::contains_bag(const input_type &input_, std::string color) {
  if (input_.count(color)) {
    for (auto &c : input_.at(color)) {
      if (c.first == "shiny gold") {
        return true;
      } else if (contains_bag(input_, c.first)) {
        return true;
      }
    }
  }
  return false;
}

std::string day7::solve_part1(const input_type &input_) {
  int n_colors(0);
  for (auto &inp : input_) {
    n_colors += (contains_bag(input_, inp.first)) ? 1 : 0;
  }
  return std::to_string(n_colors);
}

// Implement part two solution
int day7::bags_in_bag(const input_type &bag, std::string color, int n_bags) {
  if (!bag.count(color)) {
    return n_bags;
  }
  int bags_in_color(0);
  for (auto &i : bag.at(color)) {
    bags_in_color += n_bags * bags_in_bag(bag, i.first, i.second);
  }
  return n_bags + bags_in_color;
}
std::string day7::solve_part2(const input_type &input_) {
  return std::to_string(bags_in_bag(input_, "shiny gold", 1) - 1);
}
